README.md
# Hive Authentication Client (HAC)

Hive Authentication Client (HAC) is a password less users authentication, sign and broadcast transactions for the [HIVE Blockchain](https://hive.io/) through the [Hive Authentication Services (HAS)](https://docs.hiveauth.com/) or [Hive Keychain Browser Extension (KBE)](https://github.com/stoodkev/hive-keychain). 

Authentications and transactions are performed by the wallet supporting the HAS protocol or the Hive Keychain Browser Extension without communicating the private key, which thus remains secure. The results of each operation are sent by subscription (RxJS).

Hive Authentication Client (HAC) manages previous connections as well through encrypted data in the localStorage. 

**Hive Authentication Services (HAS)**

Manage the WebSocket connection to the HAS and all communication with it. Support reconnection to another HAS WebSocket server in case of failure.

**Hive Keychain Browser Extension  (KBE)**

Manage the communication with the Hive Keychain Browser Extension  if present

## Getting Started

### Installation

#### Using npm:

```bash
$  npm install @mintrawa/hive-auth-client --save
```

#### Using browser:

```html
<script type="module" src="./dist/hive-auth-client.min.js"></script>
```

#### From the source:

clone the repository

```bash
$  git clone https://github.com/Mintrawa/hive-auth-client.git
```

Build for NodeJS/Angular... in `"./lib"` directory

```bash
$  npm run build
```

Build for browser (minified) in `"./dist"` directory

```bash
$  npm run build-browser
```

## Initialisation

Initialise the Hive Authentication Client with **params**

```ts
import { HiveAuthClient } from "@mintrawa/hive-auth-client"

/** Init the Hive Authentication Client */
HiveAuthClient(["HAS server"], { debug: true })
```

or without **params** just

```ts
import { HiveAuthClient } from "@mintrawa/hive-auth-client"

/** Init the Hive Authentication Client */
HiveAuthClient()
```

**Params**
- `has` (optional): array of HAS server (default: ["wss://hive-auth.arcange.eu"])
- `options` (optional):
    - `debug` (boolean): if true activate console.log (default: false)
    - `delay` (number): delay in ms before to check the avaibility of Hive Keychain (default: 200)

*On Angular app a `delay` is needed before the avaibility of a browser extension*

## Hive Authentication Client Message subscription

All returns of information from ***HAC***, ***HAS*** or ***KBE*** are managed by RxJS messages subscription.

### hacMsg.subscribe()

Use of RxJS to send the result of user authentication and transaction signing and broadcasting.

#### Usage

```ts
import { hacMsg } from "@mintrawa/hive-auth-client"

/** Subscribe results */
hacMsg.subscribe({
  next: async message => {
    ...
  },
  error: error => {
    ...
  }
})
```

#### message

```ts
{
  "type": string
  "msg"?: {
    "result": {
      ...
    } 
  },
  "error"?: {
    "msg": string
  }
}
```

## Functions

### Main

#### hacGetConnectionStatus()

Retrieve the status of keychain and the connection status to the Hive Authentication Services (HAS)

##### Usage

```ts
import { hacGetConnectionStatus } from "@mintrawa/hive-auth-client"

/** [HAS] Get the status of the connection */
hacGetConnectionStatus()
```

##### Result
```ts
{
  "keychain": boolean
  "has": {
    "status": "connected"|"disconnected"
    "ping_rate": number
    "protocol": number
    "server": string
    "socketid": string
    "timeout": number
    "version": string
  }
}
```

#### hacGetAccounts()

Get the array of all account known for this browser or of one specific account if specified

##### Params
- `account` (optional): HIVE username to retrieve
- `pwd` (optional): password to decrypt the data if it's the first call

##### Usage

```ts
import { hacGetAccounts } from "@mintrawa/hive-auth-client"

/**
 * [HAC] Get all or one specific connection info
 * @param { string } [account] - Account to retrieve
 * @param { string } [pwd] - Password to use to decrypt localStorage
 * @returns HAC_PREVIOUS_CONNECTION[]
 */
hacGetAccounts()
```

##### Result
```ts
[{
  account: string
  has?: {
    auth_key:   string
    has_token:  string
    has_expire: number
    has_server: string
  },
  hkc: boolean,
  challenge: {
    value: string
    signature: string
  }
}]
```

#### hacRemoveAccount()

Remove an account from the storage

##### Params

- account: HIVE username to remove 

##### Usage

```ts
import { hacRemoveAccount } from "@mintrawa/hive-auth-client"

/**
 * [HAC] Remove an account
 * @param { string } account 
 * @returns boolean
 */
hacRemoveAccount(mintrawa)
```

Return `true` or `false`

#### hacCheckPwd()

Check the validity of the string password used to encrypt the localStorage

##### Params
 
- pwd: string to encrypt the localStorage

##### Usage

```ts
import { hacCheckPwd } from "@mintrawa/hive-auth-client"

/**
 * [HAC] Check password
 * @param { string } pwd
 * @returns boolean
 */
hacCheckPwd(mintrawa)
```

Return `true` or `false`

#### hacUserAuth()

Authentication of the HIVE user through the HAS or KBE

##### Params
- `account`: HIVE username to connect
- `app`: HAS_APP<sup>*</sup> information
- `pwd`: string to encrypt the localStorage
- `challenge`: object with private `key_type` ("active"|"posting") to use and string to sign
- `m` (optional): module to use "has"|"keychain" (default: "has")

HAS_APP<sup>*</sup> need to be on the form of:
```
{
  "name": string = short name of the app (ex: "peakd")
  "description"?: string = description of the app (ex: "Peakd for Hive"),
  "icon"?: string = URL to retrieve the application icon (ex: "https://peakd.com/logo.png")
}
```

##### Usage

```ts
import { hacUserAuth } from "@mintrawa/hive-auth-client"

/**
 * HAC User Authentication
 * @param { string }     account - Hive username to connect
 * @param { HAS_APP }    app - App
 * @param { string }     pwd - Password to use to encrypt localStorage
 * @param { string }     challenge - String to sign with Hive User private key 
 * @param { HAC_MODULE } [m] - Module to use (has, keychain)
 * @returns void
 */
hacUserAuth("mintrawa", { name: 'HACtutorial' }, "MyPa$$w0rd", { key_type: 'active', value: "MyCha11en6e" })
```

##### Result through subscription (RxJS)
```ts
{
  type: "authentication"
  msg?: {
    status: "authentified"|"rejected"
    data?: {
      challenge:   string
      has_token?:  string
      has_expire?: number
      has_server?: string
    } 
  },
  error?: {
    msg: string
  }
}
```

### Operations

#### hacManualTransaction()

Send a Manual Transaction

##### Params
- `key_type`: "owner"|"active"|"posting"|"memo"
- `op`: Operation recognize by the HIVE Blockchain, more info: [Hive Developer Portal](https://developers.hive.io/apidefinitions/#apidefinitions-broadcast-ops)

##### Usage

```ts
import { hacManualTransaction } from "@mintrawa/hive-auth-client"

/**
 * HAC Send a Manual Transaction
 * @param { "owner"|"active"|"posting" } key_type 
 * @param { OPERATION } op 
 */
hacManualTransaction("active", [ "claim_account", { fee: "0.000 HIVE", creator: "mintrawa", extensions: [] } ])
```

##### Result through subscription (RxJS)
```ts
{
  type: "tx_result"
  msg?: {
    status:     "accepted"|"rejected"|"signature"|"error"
    uuid?:      string
    broadcast?: boolean
    data?:      unknown
  },
  error?: {
    msg: string
  }
}
```

### Operations specifics

In order to save time on the learning curve of the HIVE blockchain, some operations exist in short version

- hacFollowing(following: string, follow: boolean)
- hacVote(author: string, permlink: string, weight: number)
- hacTransfer(to: string, amount: string, currency: "HIVE"|"HBD", memo: string)
- hacTransferToVesting(to: string, amount: string)
- hacWithdrawVesting(vesting_shares: string)
- hacDelegation(delegatee: string, vesting_shares: string)
- hacConvert(amount: string, currency: "HIVE"|"HBD")
- hacWitnessVote(witness: string, approve: boolean)
- hacWitnessProxy(proxy: string)

## Examples

https://stackblitz.com/edit/angular-ivy-ew73hs

https://github.com/Mintrawa/hac-tutorial

## Contributing

Pull requests for new features, bug fixes, and suggestions are welcome!

## License

Copyright (C) 2021  @mintrawa (https://hive.blog/@mintrawa)

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.8](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.7...v0.0.8) (2022-01-30)


### Bug Fixes

* **keychainsignbuffer:** keychainSignBuffer error when switch account ([634dd42](https://github.com/Mintrawa/hive-auth-client/commit/634dd42d8154cb21503ffe956753fc34acb6c005))

### [0.0.7](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.6...v0.0.7) (2022-01-30)


### Bug Fixes

* **switch:** fix switch account, add expire check, refactor hacRemoveAccount, update README ([2dd4c03](https://github.com/Mintrawa/hive-auth-client/commit/2dd4c0306a5bcdf8a3eb08353451ea0bc5f817cf))

### [0.0.6](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.5...v0.0.6) (2022-01-29)


### Features

* **adding:** adding functions hacRemoveAccount, hacCheckPassword ([3eb724e](https://github.com/Mintrawa/hive-auth-client/commit/3eb724e824c30f8a64c2bd0a2771667168c94504))
* **challenge:** add challenge to HAC_PREVIOUS_CONNECTION ([7053066](https://github.com/Mintrawa/hive-auth-client/commit/705306696c17c8faf8801b374b58b591ae929936))

### [0.0.5](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.4...v0.0.5) (2022-01-23)


### Features

* **hacgetconnectionstatus:** depreciate  hasGetConnectionStatus in favor of hacGetConnectionStatus ([05bb714](https://github.com/Mintrawa/hive-auth-client/commit/05bb714ab86c9d8f3c6c70d06c4d857152d0c335))
* **hastry:** limit to 10 the attempt to connect to has server ([45f980f](https://github.com/Mintrawa/hive-auth-client/commit/45f980f5a7f6b9fb1ee9f4ae3846e48561154f2c))


### Bug Fixes

* **challenge:** rename chalenge to challenge + replace ?. by if ([37d9626](https://github.com/Mintrawa/hive-auth-client/commit/37d96269715a64e3092e7ab81b0daab7e018d3ab))
* **hacgetaccounts:** test the return of hasGetAccount() ([4691aab](https://github.com/Mintrawa/hive-auth-client/commit/4691aab7ec13ebbd66133248092e188eb4099114))

### [0.0.4](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.3...v0.0.4) (2022-01-19)


### Features

* **hacclaimaccount:** hAC add hacClaimAccount specific operation function ([200d682](https://github.com/Mintrawa/hive-auth-client/commit/200d682fa13753b1b60200d705bcfc29cc35a318))

### [0.0.3](https://github.com/Mintrawa/hive-auth-client/compare/v0.0.2...v0.0.3) (2022-01-19)


### Bug Fixes

* **hacmodule:** replace "module" by "hacModule" ([b6ccf56](https://github.com/Mintrawa/hive-auth-client/commit/b6ccf564832b519dbea5ad359e3cc6f1f44775cd))

### 0.0.2 (2022-01-19)


### Bug Fixes

* **uuid:** add uuid package.json ([215304b](https://github.com/Mintrawa/hive-auth-client/commit/215304b1bfd0664e772eefdd200114e02d935662))


src/index.ts
/** RXJS */
import { Subject } from 'rxjs'

/** Encrytion */
import CryptoJS   from 'crypto-js'
import AES        from 'crypto-js/aes'

/** Hive Authentication Client */
import {
  HAC_MODULE, HAC_STATUS, HAC_PREVIOUS_CONNECTION, HAC_MESSAGE
} from './helpers/hac'

/** Hive Authentication Service */
import { 
  HiveAuthService, hasGetConnectionStatus, hasSendAuthReq, hasSetAccount, hasGetAccount, sendSignReq
} from './has'
import { 
  HAS_APP
} from './helpers/has'

/** Hive Keychain */
import { keychainCheck, keychain, keychainSignBuffer, keychainBroadcast } from './keychain'

/** HIVE */
import { CUSTOM_JSON, FOLLOWING } from './helpers/hive/custom_json'
import { 
  OPERATION, CLAIM_ACCOUNT, VOTE, ACCOUNT_WITNESS_VOTE, ACCOUNT_WITNESS_PROXY, 
  TRANSFER, TRANSFER_TO_VESTING, WITHDRAW_VESTING, DELEGATE_VESTING_SHARES, CONVERT 
} from './helpers/hive'

/** Internal fonction */
const firstCharUpper = (value: string): string => {
  return value.charAt(0).toUpperCase() + value.slice(1)
}

/**
 * 
 * HIVE AUTH CLIENT
 * 
 */

/** Default value */
let keychainDelay = 200
let hacModule: HAC_MODULE = "has"

/** HAC accounts history */
let hacAccounts: HAC_PREVIOUS_CONNECTION[] = []

let hacPwd: string
let username: string

/**
 * Hive Auth Client RxJS messaging
 */
export const hacMsg = new Subject<HAC_MESSAGE>()

/**
 * [HAC] Main Function => Hive Auth Client 
 * @param { string[] } [ hasServer ] - list of websocket url of HAS server
 * @param { debug?:boolean, delay?: number } [ options ] - options
 * @returns void
 */
export const HiveAuthClient = (hasServer?:string[], options?: { debug?:boolean, delay?: number }): void => {
  if(options) {
    sessionStorage.removeItem("hasmode")
    if(options.debug) sessionStorage.setItem("hasmode", "debug")
    if(options.delay) keychainDelay = options.delay
  }

  /** Start HiveAuthService */
  hasServer? HiveAuthService(hasServer) : HiveAuthService()

  /** Check Hive Keychain browser extension */
  keychainCheck(keychainDelay)
}

export const hacGetConnectionStatus = (): HAC_STATUS => {
  const status:HAC_STATUS = {
    keychain,
    has: hasGetConnectionStatus()
  }
  return status
}

/**
 * [HAC] Get all or one specific connection info
 * @param { string } [account] - Account to retrieve
 * @param { string } [pwd] - Password to use to decrypt localStorage
 * @returns HAC_PREVIOUS_CONNECTION[]
 */
export const hacGetAccounts = (account?: string, pwd?: string): HAC_PREVIOUS_CONNECTION[] => {
  try {
    if(pwd && typeof(pwd) !== "string") throw new Error("Password need to be a string")
    if(!pwd && typeof(hacPwd) !== "string") throw new Error("No Password yet")
    if(account && typeof(account) !== "string") throw new Error("Account is not a valid string")

    /** if password OK */
    if(hacCheckPwd(pwd ? pwd : hacPwd)) {
      if(pwd) hacPwd = pwd

      /** Decrypt Accounts */
      const a = localStorage.getItem('hac') ? localStorage.getItem('hac') : undefined
      hacAccounts = a ? JSON.parse(AES.decrypt(a.substring(64), hacPwd).toString(CryptoJS.enc.Utf8)) : []

      /** Check expired for account via HAS and remove them */
      for (const [i, acc] of hacAccounts.entries()) {
        if(acc.has && acc.has.has_expire && acc.has.has_expire < Date.now()) hacAccounts.splice(i, 1)
      }

      const enc = AES.encrypt(JSON.stringify(hacAccounts), hacPwd).toString()
      const hmac = CryptoJS.HmacSHA256(enc, CryptoJS.SHA256(hacPwd)).toString()
      localStorage.setItem('hac', hmac+enc)

      if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Accounts]', 'color: deeppink', hacAccounts)

      /** If search specific account */
      if(account) {
        const lycos = hacAccounts.find(a => a.account === account)
        if(lycos) {
          hasSetAccount(lycos)
          username = lycos.account
        }
        if(lycos && lycos.hkc) {
          hacModule = "keychain"
        } else {
          hacModule = "has"
        }
        return hasGetAccount() ? [hasGetAccount()] : []
      /** Return all accounts */
      } else {
        return hacAccounts
      }
    } else {
      localStorage.removeItem('hac')
      return []
    }
  } catch (e) {
    if(sessionStorage.getItem("hasmode")) console.error(e)
    return []
  }
}

/**
 * [HAC] Add or Update account
 * @param { HAC_PREVIOUS_CONNECTION } account 
 * @returns void
 */
export const hacAddAccount = (account: HAC_PREVIOUS_CONNECTION): void => {
  if(typeof(hacPwd) !== "string") throw new Error("No Password yet")
  username = account.account
  hasSetAccount(account)
  const lycos = hacAccounts.findIndex(a => a.account === account.account)

  if(lycos > -1) {
    hacAccounts[lycos] = account
  } else {
    hacAccounts.push(account)
  }

  const enc = AES.encrypt(JSON.stringify(hacAccounts), hacPwd).toString()
  const hmac = CryptoJS.HmacSHA256(enc, CryptoJS.SHA256(hacPwd)).toString()
  localStorage.setItem('hac', hmac+enc)
}

/**
 * [HAC] Remove an account
 * @param { string } account 
 * @returns boolean
 */
 export const hacRemoveAccount = (account: string): boolean => {
  try {
    if(account && typeof(account) !== "string") throw new Error("Account is not a valid string")
    if(hacAccounts.length === 0) throw new Error("No account in array")

    /** Search account index in array */
    const lycos = hacAccounts.findIndex(a => a.account === account)
    hacAccounts.splice(lycos, 1)

    /** if no account in array delete localStorage */
    if(hacAccounts.length === 0) {
      localStorage.removeItem('hac')
    } else {
      const enc = AES.encrypt(JSON.stringify(hacAccounts), hacPwd).toString()
      const hmac = CryptoJS.HmacSHA256(enc, CryptoJS.SHA256(hacPwd)).toString()
      localStorage.setItem('hac', hmac+enc)
    }
    return true   
  } catch (e) {
    if(sessionStorage.getItem("hasmode")) console.error(e)
    throw new Error("Something went wrong when tryin to remove the account")
  }
}

/**
 * [HAC] Check password
 * @param { string } pwd
 * @returns boolean
 */
export const hacCheckPwd = (pwd: string): boolean => {
  try {
    if(pwd && typeof(pwd) !== "string") throw new Error("Password need to be a string")
  
    /** Retrieve accounts stored */
    const a = localStorage.getItem('hac') ? localStorage.getItem('hac') : undefined
    if(a) {
      /** Check HMAC */
      const hmac = a.substring(0, 64)
      const enc  = a.substring(64)
      const decryptedhmac = CryptoJS.HmacSHA256(enc, CryptoJS.SHA256(pwd ? pwd : hacPwd)).toString()
      /** HMAC not match the reference HMAC */
      if(decryptedhmac !== hmac) {
        return false
      /** Both HMAC match */
      } else {
        return true
      }
    /** No accounts stored yet */
    } else {
      throw new Error("No accounts stored yet")
    }
  /** Something went wrong */
  } catch (e) {
    if(sessionStorage.getItem("hasmode")) console.error(e)
    throw new Error("Something went wrong with the password check")
  }
}

/**
 * HAC User Authentication
 * @param { string }     account - Hive User to connect
 * @param { HAS_APP }    app - App
 * @param { string }     pwd - Password to use to encrypt localStorage
 * @param { string }     challenge - String to sign with Hive User private key 
 * @param { HAC_MODULE } [m] - Module to use (has, keychain)
 * @returns void
 */
export const hacUserAuth = (account: string, app: HAS_APP, pwd: string, challenge: { key_type: "active"|"posting", value: string }, m?: HAC_MODULE): void => {
  try {
    if(typeof(account) !== "string") throw new Error("Account need to be a string")
    if(typeof(pwd) !== "string") throw new Error("Password need to be a string")
    if(typeof(challenge.value) !== "string") throw new Error("Challenge value need to be a string")
    if(!challenge.key_type || (challenge.key_type !== "active" && challenge.key_type !== "posting")) throw new Error('Key value need to be "posting" or "active"')

    /** if previous */
    if(localStorage.getItem('hac')) {
      /** Retrieve account if known */
      const a = hacGetAccounts(account, pwd)
      if(a.length === 1 && a[0].hkc) {
        hacModule = "keychain"
      } else {
        hacModule = "has"
      }
    } else {
      if(pwd) hacPwd = pwd
    }

    /** If force module */
    if(m) hacModule = m

    /** Auth by Hive Keychain */
    if(hacModule === "keychain") keychainSignBuffer(account, challenge.value, firstCharUpper(challenge.key_type) as "Owner"|"Active"|"Posting"|"Memo", keychainDelay)

    /** Auth by HAS */
    if(hacModule === "has") hasSendAuthReq(account, app, challenge)
  } catch (e) {
    hacMsg.next({ type: "authentication", error: { msg: e instanceof Error ? e.message : 'error' } })
  }
}

/*****************************
 * 
 * OPERATIONS
 * 
 */

/**
 * HAC Send a Manual Transaction
 * @param { "owner"|"active"|"posting" } key_type 
 * @param { OPERATION } op 
 */
export const hacManualTransaction = (key_type: "owner"|"active"|"posting"|"memo", op: OPERATION ): void => {
  if(typeof(key_type) !== "string" && !["owner","active","posting"].includes(key_type)) throw new Error('Not a valid key_type')
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Following]', 'color: deeppink', op)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [op], firstCharUpper(key_type) as "Owner"|"Active"|"Posting"|"Memo", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type, ops: [op], broadcast: true })
}

/**
 * HAC Claim a free account
 */
export const hacClaimAccount = (): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")

  const claimAccount: CLAIM_ACCOUNT = [ "claim_account", { fee: "0.000 HIVE", creator: username, extensions: [] } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Following]', 'color: deeppink', claimAccount)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [claimAccount], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [claimAccount], broadcast: true })
}

/**
 * Follow/UnFollow a user
 * @param { string } account 
 * @param { string } following 
 * @param { boolean } follow 
 */
export const hacFollowing = (following: string, follow: boolean): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")

  const json: FOLLOWING = [ "follow", { follower: username, following, what: follow ? [ "blog" ] : [] } ]
  const custom_json: CUSTOM_JSON = [ "custom_json", { id: 'follow', json: JSON.stringify(json), required_auths: [], required_posting_auths: [ username ] } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Following]', 'color: deeppink', custom_json)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [custom_json], "Posting", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "posting", ops: [custom_json], broadcast: true })
}

/**
 * Vote/Downvote a user post
 * @param { string } author 
 * @param { string } permlink 
 * @param { number } weight 
 */
export const hacVote = (author: string, permlink: string, weight: number): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(permlink) !== "string") throw new Error("Permlink Error")
  if(weight < -100 || weight > 100) throw new Error("weight need to be between -100 and 100")

  const vote: VOTE = [ "vote", { voter: username, author, permlink, weight: weight * 100 } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Vote]', 'color: deeppink', vote)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [vote], "Posting", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "posting", ops: [vote], broadcast: true })
}

/*****************************
 * 
 * OPERATIONS WALLET
 * 
 */

/**
 * Transfer HIVE or HBD to another user
 * @param to 
 * @param amount 
 * @param currency 
 * @param memo 
 */
export const hacTransfer = (to: string, amount: string, currency: "HIVE"|"HBD", memo: string): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(to) !== "string") throw new Error("recipient must be a string")
  if(typeof(amount) !== "string") throw new Error("amount must be a string")
  if(currency !== "HIVE" && currency !== "HBD") throw new Error("currency must be HIVE or HBD")
  if(typeof(memo) !== "string") throw new Error("memo must be a string")

  const transfer: TRANSFER = [ "transfer", { from: username, to, amount: amount.concat(" ", currency), memo } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Transfer]', 'color: deeppink', transfer)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [transfer], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [transfer], broadcast: true })
}

/**
 * Transfer to Vesting (Power UP)
 * @param { string } to 
 * @param { string } amount 
 */
export const hacTransferToVesting = (to: string, amount: string): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(to) !== "string") throw new Error("recipient must be a string")
  if(typeof(amount) !== "string") throw new Error("amount must be a string")

  const transferToVesting: TRANSFER_TO_VESTING = [ "transfer_to_vesting", { from: username, to, amount: amount+" HIVE" } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Transfer To Vesting]', 'color: deeppink', transferToVesting)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [transferToVesting], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [transferToVesting], broadcast: true })
}

/**
 * Withdraw from Vesting (Power Down)
 * @param { string } vesting_shares 
 */
export const hacWithdrawVesting = (vesting_shares: string): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(vesting_shares) !== "string") throw new Error("amount must be a string")

  const withdrawVesting: WITHDRAW_VESTING = [ "withdraw_vesting", { account: username, vesting_shares } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Withdraw Vesting]', 'color: deeppink', withdrawVesting)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [withdrawVesting], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [withdrawVesting], broadcast: true })
}

/**
 * Delegate HIVE POWER to a user
 * @param { string } delegatee 
 * @param { string } amount 
 */
export const hacDelegation = (delegatee: string, vesting_shares: string): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(delegatee) !== "string") throw new Error("Delegatee must be a string")
  if(typeof(vesting_shares) !== "string") throw new Error("amount must be a string")

  const delegation: DELEGATE_VESTING_SHARES = [ "delegate_vesting_shares", { delegator: username, delegatee, vesting_shares } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Delegation]', 'color: deeppink', delegation)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [delegation], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [delegation], broadcast: true })
}

/**
 * Convert HBD => HIVE or HIVE => HBD
 * @param { string } amount
 * @param { "HIVE"|"HBD" } currency
 */
export const hacConvert = (amount: string, currency: "HIVE"|"HBD"): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(amount) !== "string") throw new Error("amount must be a string")
  //create random number for requestid paramter
  const requestid = Math.floor(Math.random() * 10000000)
  const convert: CONVERT = [ currency === "HBD" ? "convert" : "collateralized_convert", { owner: username, requestid, amount: amount.concat(" ", currency) } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Convert]', 'color: deeppink', convert)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [convert], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [convert], broadcast: true })
}

/*****************************
 * 
 * OPERATIONS WITNESS
 * 
 */

/**
 * Approve/Disapprove a witness
 * @param { string } witness 
 * @param { boolean } approve 
 */
export const hacWitnessVote = (witness: string, approve: boolean): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(witness) !== "string") throw new Error("Witness Error")
  if(typeof(approve) !== "boolean") throw new Error("Approve Error")

  const witnessVote: ACCOUNT_WITNESS_VOTE = [ "account_witness_vote", { account: username, witness, approve } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Vote Witness]', 'color: deeppink', witnessVote)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [witnessVote], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [witnessVote], broadcast: true })
}

/**
 * Set a proxy as Witness
 * @param { string } proxy 
 */
export const hacWitnessProxy = (proxy: string): void => {
  if(typeof(username) !== "string") throw new Error("No user connected yet")
  if(typeof(proxy) !== "string") throw new Error("Proxy Error")

  const witnessProxy: ACCOUNT_WITNESS_PROXY = [ "account_witness_proxy", { account: username, proxy } ]
  if(sessionStorage.getItem("hasmode"))  console.log('%c[HAC Witness Proxy]', 'color: deeppink', witnessProxy)
  /** Keychain */
  if(hacModule === "keychain") keychainBroadcast(username, [witnessProxy], "Active", keychainDelay)
  /** HAS */
  if(hacModule === "has") sendSignReq(username, { key_type: "active", ops: [witnessProxy], broadcast: true })
}

export { hasGetConnectionStatus } from "./has"
export {
  HAC_PREVIOUS_CONNECTION,
  HAC_KEYCHAIN_STATUS,
  HAC_MSG_QR_CODE,
  HAC_MSG_AUTHENTICATION,
  HAC_MSG_SIGN_WAIT,
  HAC_MSG_TX_RESULT
} from './helpers/hac'